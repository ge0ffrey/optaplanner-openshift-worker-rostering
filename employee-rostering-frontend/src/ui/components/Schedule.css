/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
 :root {
    --grid-header-column-width: 150px;
    --grid-soft-line-interval: var(--grid-unit-size);
    --grid-hard-line-interval: calc(var(--grid-soft-line-interval) * 6);
    --grid-unit-size: 100px;
    --grid-row-size: 50px;
  }

 .viewport-overlay {
    position: fixed;
    top: 0;
    left: 0;
    width: 100vw;
    height: 100vh;
    background-color: rgba(0,0,0,0.2);
    z-index: 500;
}

.schedule {
    display: grid;
    overflow: scroll;
}

.date-time-corner {
	position: sticky;
    left: 0;
    grid-column-start: 1;
    grid-column-end: 1;
    background-color: white;
    z-index: 20;
    border: 1px solid;
    width: var(--grid-header-column-width);
}

.date-time-lane {
    position: sticky;
    box-sizing: border-box;
    display: flex;
    top: 0;
	z-index: 20;
}

.date-time-container {
    display: grid;
    background-color: white;
    border-bottom: 1px solid black;
    grid-template-rows: min-content max-content;
    grid-auto-columns: var(--grid-unit-size);
}

.date-tick {
    font-size: 1.0em;
    height: 25px;
    border: 1px solid lightgray;
    border-right: none;
    opacity: 0.7;
    text-overflow: ellipsis;
    overflow: hidden;
    text-align: center;
    grid-row: 1;
}

.date-change-tick {
   border-left: 1px solid lightgray;
   grid-row: 2;
}

.time-tick {
   text-indent: -1.25em;
   font-size: 0.8em;
   transform-origin: left;
   opacity: 0.7;
   padding-top: 100%;
   overflow: visible;
   white-space: nowrap;
   transform: translateX(-10px) rotateZ(-60deg);
   grid-row: 2;
}

.lane-container {
    display: flex;
    box-sizing: border-box;
}

.lane-title {
    position: sticky;
    display: block;
    text-overflow: ellipsis;
    left: 0;
    padding-left: 10px;
    background-color: white;
    border: 1px solid black;
    border-top: none;
    width: var(--grid-header-column-width);
    z-index: 19;
}

.lane-content {
    position: relative;
	display: inline-grid;
	grid-auto-flow: row dense;
	grid-auto-columns: var(--grid-unit-size);
	grid-auto-rows: var(--grid-row-size);;
	border-bottom: 1px solid black;
	border-right: 1px solid black;
	background-size: var(--grid-soft-line-interval) var(--grid-soft-line-interval),
	                 var(--grid-hard-line-interval) var(--grid-hard-line-interval),
	                 50px 50px,
	                 50px 50px;
	background-image: linear-gradient(to right, rgba(0, 0, 0, 0.5) 1px, transparent 1px),
	                  linear-gradient(to right, rgba(0, 0, 0, 0.8) 1px, transparent 1px),
	                  repeating-linear-gradient(to right, #ffffff, #ffffff 1px, transparent 10px, transparent 1px),
	                  repeating-linear-gradient(to bottom, rgba(0, 0, 0, 0.2), rgba(0, 0, 0, 0.2) 1px, transparent 1px, transparent 50px);
}

.blob, .spanning-blob {
	width: calc(100% - 2px);
    height: calc(100% - 2px);
    z-index: 10;
    color: black;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    background-color: rgba(166, 214, 181, 0.49);
    font-size: 1em !important;
    padding: 26px 2px 2px 2px;
    padding-top: 10px !important;
    border-radius: 2px;
    opacity: 0.8;
}

.blob {
	width: calc(100% - 2px);
    height: calc(100% - 2px);
}

.score {
	animation-name: score-highlight;
	animation-play-state: paused;
	animation-duration: 100s;
}

/* TODO: Get UX input on colors */
@keyframes score-highlight {
  /* Negative Hard */
  0% {
    background-color: rgba(139, 0, 0, 0.49);
  }
  
  /* Negative Medium */ 
  15% {
    background-color: rgba(245, 193, 46, 0.49);
  }
  
  /* Negative Soft */
  30% {
    background-color: rgba(209, 209, 209, 0.49);
  }
  
  /* Zero Score */
  50% {
    background-color: rgba(207, 231, 205, 0.49);
  }
  
  /* Positive Soft */
  70% {
    background-color: rgba(63, 156, 53, 0.49);
  }
  
  /* Positive Medium */
  85% {
    background-color: rgba(0, 185, 228, 0.49);
  }
  
  /* Positive Hard */
  100% {
     background-color: rgba(112, 63, 236, 0.49);
  }
} 

.spanning-blob {
	position: absolute;
	width: 100%;
    height: 100%;
}

.spanning-blob.selected, .blob.selected {
    border: 2px solid #e3ffec !important;
    border-radius: 2px;
    z-index: 1002;
    box-shadow: inset 0px 0px 20px 6px white;
}

.spanning-blob:hover, .blob:hover {
    opacity: 1;
}

.spanning-blob:active, .blob:active {
    opacity: 1;
}

.spanning-blob.historic, .blob.historic {
}

.spanning-blob.published, .blob.published {
    border: 1px solid black;
}

.spanning-blob.draft, .blob.draft {
    border: 1px dashed black;
}

.unassigned {
    background-color: rgba(245, 193, 46, 0.49) !important;
}

.show-on-hover {
    display: none;
}

*:not(.selected):hover > .show-on-hover {
    display: grid;
}

*:not(.selected) > .show-on-hover:hover {
    display: grid;
}