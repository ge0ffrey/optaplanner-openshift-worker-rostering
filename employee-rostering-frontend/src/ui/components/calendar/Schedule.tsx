/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as React from 'react';
import { Calendar, momentLocalizer, EventProps } from 'react-big-calendar'
import moment from 'moment';
import EventWrapper from './EventWrapper';

import 'react-big-calendar/lib/css/react-big-calendar.css';
import './ReactBigCalendarOverrides.css';

export interface Props<T extends object> {
  startDate: Date;
  endDate: Date;
  events: T[];
  showAllDayCell?: boolean;
  dateFormat?: (date: Date) => string;
  startAccessor: (event: T) => Date;
  endAccessor: (event: T) => Date;
  titleAccessor: (event: T) => string;
  addEvent: (start: Date, end: Date) => void;
  eventStyle: (event: T) => { style?: React.CSSProperties; className?: string };
  wrapperStyle: (event: T) => { style?: React.CSSProperties; className?: string; props?: {} };
  dayStyle: (date: Date) => { style?: React.CSSProperties; className?: string };
  popoverHeader: (event: T) => React.ReactNode;
  popoverBody: (event: T) => React.ReactNode;
  eventComponent: (props: React.PropsWithChildren<EventProps<T>>) => React.ReactNode;
}

export function isDay(start: Date, end: Date) {
  return start.getHours() === 0 && start.getMinutes() === 0 &&
    end.getHours() === 0 && end.getMinutes() === 0
}

const localizer = momentLocalizer(moment);
export default function Schedule<T extends object>(props: Props<T>): React.ReactElement<Props<T>> {
  const length = Math.ceil(moment.duration(moment(props.startDate).to(moment(props.endDate))).asDays());
  return (
    <div style={{
      height: "calc(100% - 20px)"
    }}
    >
      <Calendar
        className={(props.showAllDayCell)? undefined : "rbc-no-allday-cell"}
        date={props.startDate}
        length={length}
        localizer={localizer}
        events={props.events}
        titleAccessor={props.titleAccessor}
        allDayAccessor={event => props.showAllDayCell?
          isDay(props.startAccessor(event), props.endAccessor(event)) : false}
        startAccessor={props.startAccessor}
        endAccessor={props.endAccessor}
        toolbar={false}
        view="week"
        views={["week"]}
        formats={props.dateFormat? {
          dayFormat: props.dateFormat
        } : undefined}
        onSelectSlot={(slotInfo: { start: string|Date; end: string|Date;
          action: "select"|"click"|"doubleClick"; }) => {
          if (slotInfo.action === "select" || slotInfo.action === "click") {
            if (isDay(moment(slotInfo.start).toDate(), moment(slotInfo.end).toDate())) {
              props.addEvent(moment(slotInfo.start).toDate(), moment(slotInfo.end).add(1, "day").toDate());
            }
            else {
              props.addEvent(moment(slotInfo.start).toDate(), moment(slotInfo.end).toDate());
            }
          }
        }
        }
        onView={() => {}}
        onNavigate={() => {}}
        timeslots={4}
        eventPropGetter={props.eventStyle}
        dayPropGetter={props.dayStyle}
        selectable
        showMultiDayTimes
        components={{
          eventWrapper: (wrapperProps) => {
            const style = props.wrapperStyle(wrapperProps.event);
            return EventWrapper({
              ...wrapperProps,
              style: {
                ...wrapperProps.style,
                ...((style.style)? style.style : {}) 
              },
              className: style.className? style.className : "",
              additionalProps: style.props? style.props : {},
              popoverHeader: props.popoverHeader(wrapperProps.event),
              popoverBody: props.popoverBody(wrapperProps.event)
            })
          },
          event: (eventProps) => props.eventComponent(eventProps) as React.ReactElement
        }}
      />
    </div>
  )
}