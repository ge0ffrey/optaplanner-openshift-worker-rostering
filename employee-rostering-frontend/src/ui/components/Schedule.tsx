/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import moment from 'moment';
import './Schedule.css';
import DomainObject from 'domain/DomainObject';
import GridLayout from 'react-grid-layout';

export interface ScheduleProps<T extends DomainObject> {
  startDate: Date;
  endDate: Date;
  minDurationInMinutes: number;
  rowTitles: string[];
  rowData: T[][];
  dataGetStartDate: (data: T) => Date;
  dataGetEndDate: (data: T) => Date;
  dataToNameMap: (data: T) => string;
}

type CSSVariable = "--grid-header-column-width" | "--grid-soft-line-interval" | "--grid-hard-line-interval" | "--grid-unit-size" | "--grid-row-size"

export interface ScheduleState {
  cssVariables: React.CSSProperties;
}

export default abstract class Schedule<T extends DomainObject> extends React.Component<ScheduleProps<T>, ScheduleState> {
  
  windowResizeListener: (e: UIEvent) => void;

  constructor(props: ScheduleProps<T>) {
    super(props);
    this.getDateTimeLane = this.getDateTimeLane.bind(this);
    this.getDataLane = this.getDataLane.bind(this);
    this.getCSSVariable = this.getCSSVariable.bind(this);
    const gridCount = moment(this.props.endDate.getTime()).diff(this.props.startDate.getTime(), "minutes") / this.props.minDurationInMinutes;
    const gridUnitSize = Math.max(((window.innerWidth - 200) / gridCount), 45);
    this.state = {
      cssVariables: {
        "--grid-unit-size": gridUnitSize + "px",
        "--grid-row-size": "50px",
        "--grid-soft-line-interval": gridUnitSize + "px",
        "--grid-hard-line-interval": gridUnitSize * (moment.duration(1, "day").asMinutes() / this.props.minDurationInMinutes) + "px"
      } as React.CSSProperties
    };

    this.windowResizeListener = (e) => {
      const currentGridCount = moment(this.props.endDate.getTime()).diff(this.props.startDate.getTime(), "minutes") / this.props.minDurationInMinutes;
      const gridUnitSize =  Math.max(((e.target as Window).outerWidth - 200) / currentGridCount, 45);
      this.setCSSVariable("--grid-unit-size", gridUnitSize + "px");
      this.setCSSVariable("--grid-soft-line-interval", gridUnitSize + "px");
      this.setCSSVariable("--grid-hard-line-interval", gridUnitSize * (moment.duration(1, "day").asMinutes() / this.props.minDurationInMinutes) + "px");
      this.forceUpdate();
    }
  }

  componentDidMount() {
    window.addEventListener("resize", this.windowResizeListener);
  }

  componentWillUnmount() {
    window.removeEventListener("resize", this.windowResizeListener);
  }

  setCSSVariable(variable: CSSVariable, value: string): void {
    this.setState((prevState) => ({
      cssVariables: {...prevState.cssVariables, [variable]: value}
    }));
  }

  getCSSVariable(variable: CSSVariable): string {
    return (this.state.cssVariables as any)[variable];
  }

  getGridPosition(start: number, end: number): React.CSSProperties {
    const snapStartPositionInGrid = Math.floor(start);
    const startPositionRemainder = start - snapStartPositionInGrid;

    const snapEndPositionInGrid = Math.floor(end);
    const endPositionRemainder = end - snapEndPositionInGrid;

    return {
      gridColumnStart: snapStartPositionInGrid + 1,
      marginLeft: "calc(" + startPositionRemainder + "* var(--grid-unit-size))",
      gridColumnEnd: snapEndPositionInGrid + 1,
      marginRight: "calc(" + endPositionRemainder + "* var(--grid-unit-size))",
    };
  }

  getDateTimeLane(durationInDays: number, gridCount: number): JSX.Element {
    const gridUnitsPerDay = (60 * 24) / this.props.minDurationInMinutes;
    return (
      <div className="date-time-lane">
        <div className="date-time-corner" />
        <div className="date-time-container">
          {
            Array.from(new Array(durationInDays)).map((v, index) => {
              const date = moment(this.props.startDate)
                .add(index, "days");
              return (
                <div
                  key={date.format("L")}
                  className="date-tick"
                  style={this.getGridPosition(index * gridUnitsPerDay, (index + 1) * gridUnitsPerDay)}
                >
                  {date.format("L")}
                </div>
              );
            })
          }
          {
            Array.from(new Array(durationInDays)).map((v, index) => {
              return (
                <div
                  key={moment(this.props.startDate).add(index, "days").toString()}
                  className="date-change-tick"
                  style={this.getGridPosition(index * gridUnitsPerDay, (index + 1) * gridUnitsPerDay)}
                />
              );
            })
          }
          {
            Array.from(new Array(gridCount)).map((v, index) => {
              const date = moment(this.props.startDate)
                .add(this.props.minDurationInMinutes * index, "minutes");
              return (
                <div
                  key={date.format("LLLL")}
                  className="time-tick"
                  style={this.getGridPosition(index, index)}
                >
                  {date.format("LT")}
                </div>
              );
            })
          }
        </div>
      </div>
    );
  }

  abstract getDataLane(title: string, row: number, gridCount: number): JSX.Element;

  render() {
    const gridCount = moment(this.props.endDate.getTime()).diff(this.props.startDate.getTime(), "minutes") / this.props.minDurationInMinutes;
    const durationInDays =  moment(this.props.endDate.getTime()).diff(this.props.startDate.getTime(), "days"); 
    return (
      <div className="schedule" style={this.state.cssVariables}>
        {this.getDateTimeLane(durationInDays, gridCount)}
        {this.props.rowTitles.map((title, row) => this.getDataLane(title, row, gridCount))}
      </div>
    );
  }
}