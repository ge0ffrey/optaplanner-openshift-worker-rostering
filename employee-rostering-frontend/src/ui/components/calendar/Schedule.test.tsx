/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as React from 'react';
import Schedule, { isDay, Props } from './Schedule';
import { shallow, ShallowWrapper } from 'enzyme';
import { CalendarProps, EventProps, EventWrapperProps } from 'react-big-calendar';

function calendarProps(wrapper: ShallowWrapper): CalendarProps<{ start: Date; end: Date; title: string }> {
  return (wrapper.find("ForwardRef").props() as CalendarProps<{ start: Date; end: Date; title: string }>);
}

function reportError(prop: string): never {
  throw new Error(prop + "  should be defined.");
}

describe('Schedule', () => {
  it('isDay should return true if both start and end are on midnight', () => {
    const startDate = new Date("2018-01-01T00:00");
    const endDate = new Date("2018-01-02T00:00");
    expect(isDay(startDate, endDate)).toEqual(true);
  });

  it('isDay should return false if either start or end are not midnight', () => {
    const startDate1 = new Date("2018-01-01T12:00");
    const endDate1 = new Date("2018-01-02T00:00");
    expect(isDay(startDate1, endDate1)).toEqual(false);

    const startDate2 = new Date("2018-01-01T00:00");
    const endDate2 = new Date("2018-01-02T12:00");
    expect(isDay(startDate2, endDate2)).toEqual(false);
  });

  it('should render correctly without an all-day cell', () => {
    const schedule = shallow(<Schedule {...props} />);
    const allDayAccessor: Function = calendarProps(schedule).allDayAccessor as Function;
    expect(allDayAccessor(props.events[0])).toEqual(false);
    expect(allDayAccessor(props.events[1])).toEqual(false);
    expect(allDayAccessor(props.events[2])).toEqual(false);
  });

  it('should render correctly with an all-day cell', () => {
    const schedule = shallow(<Schedule {...props} showAllDayCell />);
    const allDayAccessor: Function = calendarProps(schedule).allDayAccessor as Function;
    expect(allDayAccessor(props.events[0])).toEqual(false);
    expect(allDayAccessor(props.events[1])).toEqual(true);
    expect(allDayAccessor(props.events[2])).toEqual(false);
  });

  it('length should be correct', () => {
    const schedule = shallow(<Schedule {...props} />);
    const length = calendarProps(schedule).length;
    expect(length).toEqual(7);
  });

  it('title accessor should be prop', () => {
    const schedule = shallow(<Schedule {...props} />);
    const titleAccessor = calendarProps(schedule).titleAccessor;
    expect(titleAccessor).toBe(props.titleAccessor);
  });

  it('start accessor should be prop', () => {
    const schedule = shallow(<Schedule {...props} />);
    const startAccessor = calendarProps(schedule).startAccessor;
    expect(startAccessor).toBe(props.startAccessor);
  });

  it('end accessor should be prop', () => {
    const schedule = shallow(<Schedule {...props} />);
    const endAccessor = calendarProps(schedule).endAccessor;
    expect(endAccessor).toBe(props.endAccessor);
  });

  it('day format should be date format if provided', () => {
    const dateFormat = jest.fn();
    const schedule = shallow(<Schedule {...props} dateFormat={dateFormat} />);
    const { dayFormat } = calendarProps(schedule).formats || reportError("formats");
    expect(dayFormat).toBe(dateFormat);
  });
  
  it('formats should be undefined if date format not provided', () => {
    const schedule = shallow(<Schedule {...props} />);
    expect(calendarProps(schedule).formats).toBeUndefined();
  });

  it('should not create event on double click', () => {
    const schedule = shallow(<Schedule {...props} />);
    const onSelectSlot = calendarProps(schedule).onSelectSlot || reportError("onSelectSlot");
    onSelectSlot({
      start: props.events[0].start,
      end: props.events[0].end,
      slots: [],
      action: "doubleClick"
    });
    expect(props.addEvent).not.toBeCalled();
  });

  it('should create event on click', () => {
    const schedule = shallow(<Schedule {...props} />);
    const onSelectSlot = calendarProps(schedule).onSelectSlot || reportError("onSelectSlot");
    onSelectSlot({
      start: props.events[0].start,
      end: props.events[0].end,
      slots: [],
      action: "click"
    });
    expect(props.addEvent).toBeCalled();
    expect(props.addEvent).toBeCalledWith(props.events[0].start, props.events[0].end);
  });

  it('should create event on select', () => {
    const schedule = shallow(<Schedule {...props} />);
    const onSelectSlot = calendarProps(schedule).onSelectSlot || reportError("onSelectSlot");
    onSelectSlot({
      start: props.events[0].start,
      end: props.events[0].end,
      slots: [],
      action: "select"
    });
    expect(props.addEvent).toBeCalled();
    expect(props.addEvent).toBeCalledWith(props.events[0].start, props.events[0].end);
  });

  it('should create event of proper length if it span a day', () => {
    const schedule = shallow(<Schedule {...props} />);
    const onSelectSlot = calendarProps(schedule).onSelectSlot || reportError("onSelectSlot");
    onSelectSlot({
      start: props.events[1].start,
      end: props.events[1].start, // Not an error, selecting a single day yields start and end the same
      slots: [],
      action: "click"
    });
    expect(props.addEvent).toBeCalled();
    expect(props.addEvent).toBeCalledWith(props.events[1].start, props.events[1].end);
  });

  it('eventPropGetter should be prop', () => {
    const schedule = shallow(<Schedule {...props} />);
    const eventPropGetter = calendarProps(schedule).eventPropGetter;
    expect(eventPropGetter).toBe(props.eventStyle);
  });
  
  it('dayPropGetter should be prop', () => {
    const schedule = shallow(<Schedule {...props} />);
    const dayPropGetter = calendarProps(schedule).dayPropGetter;
    expect(dayPropGetter).toBe(props.dayStyle);
  });

  it('event component should defer to prop', () => {
    const schedule = shallow(<Schedule {...props} />);
    const components = calendarProps(schedule).components || reportError("components");
    const event = components.event as 
      React.FC<React.PropsWithChildren<EventProps<{ start: Date; end: Date; title: string }>>> 
      || reportError("components.event");
    const out = event(
      {
        event: props.events[0],
        ...props.events[0]
      });
    expect(props.eventComponent).toBeCalled();
    expect(props.eventComponent).toBeCalledWith({
      event: props.events[0],
      ...props.events[0]
    })
    expect(out).toEqual(props.eventComponent({
      event: props.events[0],
      ...props.events[0]
    }))
  });

  it('wrapper component should defer to prop', () => {
    const schedule = shallow(<Schedule {...props} />);
    const components = calendarProps(schedule).components || reportError("components");
    const wrapper = components.eventWrapper as 
      React.FC<React.PropsWithChildren<EventWrapperProps<{ start: Date; end: Date; title: string }>>> 
      || reportError("components.event");
    const wrapperProps: React.PropsWithChildren<EventWrapperProps<{ start: Date; end: Date; title: string }>> =
    {
      event: props.events[0],
      className: "wrapperClass",
      isRtl: false,
      getters: {},
      onClick: jest.fn(),
      onDoubleClick: jest.fn(),
      accessors: {},
      selected: false,
      label: "Label",
      continuesEarlier: false,
      continuesLater: true,
      style: {
        top: "33%",
        height: "67%",
        xOffset: 0
      }
    }
    const out = wrapper(wrapperProps);
    expect(props.wrapperStyle).toBeCalled();
    expect(props.wrapperStyle).toBeCalledWith(props.events[0]);
    expect(out).toMatchSnapshot();
  });
});

const props: Props<{ start: Date; end: Date; title: string }> = {
  startDate: new Date("2018-01-01T00:00"),
  endDate: new Date("2018-01-07T00:00"),
  events: [
    {
      title: "Event 1",
      start: new Date("2018-01-02T08:00"),
      end: new Date("2018-01-02T17:00")      
    },
    {
      title: "All Day Event ",
      start: new Date("2018-01-03T00:00"),
      end: new Date("2018-01-04T00:00")      
    },
    {
      title: "Multi-day Event",
      start: new Date("2018-01-0520:00"),
      end: new Date("2018-01-06T06:00")      
    }
  ],
  showAllDayCell: false,
  startAccessor: jest.fn(e => e.start),
  endAccessor: jest.fn(e => e.end),
  titleAccessor: jest.fn(e => e.title),
  addEvent: jest.fn(),
  eventStyle: jest.fn(() => ({ style: { color: "red" }, className: "class" })),
  wrapperStyle: jest.fn(() => ({ style: { color: "red" }, className: "class", props: { prop: true } })),
  dayStyle: jest.fn(() => ({ style: { color: "red" }, className: "class" })),
  popoverHeader: jest.fn(() => "Title"),
  popoverBody: jest.fn(() => "Body"),
  eventComponent: jest.fn(e => e.title)
}