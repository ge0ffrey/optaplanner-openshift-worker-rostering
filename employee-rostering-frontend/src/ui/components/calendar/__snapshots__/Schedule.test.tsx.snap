// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schedule wrapper component should defer to prop 1`] = `
<Popover
  appendTo={[Function]}
  aria-label=""
  bodyContent="Body"
  className=""
  closeBtnAriaLabel="Close"
  enableFlip={true}
  footerContent={null}
  headerContent="Title"
  hideOnOutsideClick={true}
  isVisible={null}
  maxWidth="calc(1.5rem + 18.75rem)"
  onHidden={[Function]}
  onHide={[Function]}
  onMount={[Function]}
  onShow={[Function]}
  onShown={[Function]}
  position="right"
  shouldClose={[Function]}
  zIndex={9999}
>
  <div
    className="class continues-next-day"
    style={
      Object {
        "backgroundColor": "transparent",
        "border": "none",
        "color": "red",
        "gridRowEnd": 101,
        "gridRowStart": 34,
        "height": "67%",
        "top": "33%",
        "xOffset": 0,
      }
    }
  />
</Popover>
`;
